File: main/src/main/scala/org/clulab/dynet/RnnLayer.scala
package org.clulab.dynet
import java.io.PrintWriter

import edu.cmu.dynet.{Expression, ExpressionVector, GruBuilder, LstmBuilder, ParameterCollection, RnnBuilder}
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.mutable.ArrayBuffer
import org.clulab.dynet.Utils._
import org.clulab.utils.Configured

/**
 * This layer applies a biLSTM over the sequence of Expressions produced by a previous layer
 * @author Mihai
 */
class RnnLayer (val parameters:ParameterCollection,
                val inputSize: Int,
                val numLayers: Int,
                val rnnStateSize: Int,
                val useHighwayConnections: Boolean,
                val rnnType: String, // "gru" or "lstm"
                val wordFwRnnBuilder:RnnBuilder,
                val wordBwRnnBuilder:RnnBuilder,
                val dropoutProb: Float) extends IntermediateLayer {

  override def forward(inputExpressions: ExpressionVector, doDropout: Boolean): ExpressionVector = {
    setRnnDropout(wordFwRnnBuilder, dropoutProb, doDropout)
    setRnnDropout(wordBwRnnBuilder, dropoutProb, doDropout)
    setRnnDropout(wordBwRnnBuilder, dropoutProb, doDropout)

    assert(inputExpressions.nonEmpty)
    val fwEmbeddings = inputExpressions
    val fwStates = Utils.transduce(fwEmbeddings, wordFwRnnBuilder)
    assert(fwStates.length == inputExpressions.length)
    val bwEmbeddings = fwEmbeddings.reverse
    val bwStates = Utils.transduce(bwEmbeddings, wordBwRnnBuilder).reverse
    assert(bwStates.length == inputExpressions.length)

    // the word state concatenates the fwd and bwd LSTM hidden states; and the input embedding if useHighwayConnections
    val states = fwStates.indices.map { i =>
      if (useHighwayConnections)
        Expression.concatenate(fwStates(i), bwStates(i), inputExpressions(i))
      else
        Expression.concatenate(fwStates(i), bwStates(i))
    }

    states
  }

  override def outDim: Int = {
    val highwaySize = if(useHighwayConnections) inputSize else 0
    2 * rnnStateSize + highwaySize
  }

  override def inDim: Int = inputSize

  override def saveX2i(printWriter: PrintWriter): Unit = {
    save(printWriter, inputSize, "inputSize")
    save(printWriter, numLayers, "numLayers")
    save(printWriter, rnnStateSize, "rnnStateSize")
    save(printWriter, if(useHighwayConnections) 1 else 0, "useHighwayConnections")
    save(printWriter, rnnType, "rnnType")
    save(printWriter, dropoutProb, "dropoutProb")
  }

  override def toString: String = {
    s"RnnLayer($rnnType, $inDim, $outDim)"
  }
}

object RnnLayer {
  val logger: Logger = LoggerFactory.getLogger(classOf[RnnLayer])

  val DEFAULT_DROPOUT_PROB: Float = Utils.DEFAULT_DROPOUT_PROBABILITY

  def load(parameters: ParameterCollection,
           x2iIterator: BufferedIterator[String]): RnnLayer = {
    //
    // load the x2i info
    //
    val byLineIntBuilder = new ByLineIntBuilder()
    val byLineFloatBuilder = new ByLineFloatBuilder()
    val byLineStringBuilder = new ByLineStringBuilder()

    val inputSize = byLineIntBuilder.build(x2iIterator, "inputSize")
    val numLayers = byLineIntBuilder.build(x2iIterator, "numLayers")
    val rnnStateSize = byLineIntBuilder.build(x2iIterator, "rnnStateSize")
    val useHighwayConnectionsAsInt = byLineIntBuilder.build(x2iIterator, "useHighwayConnections")
    val useHighwayConnections = useHighwayConnectionsAsInt == 1
    val rnnType = byLineStringBuilder.build(x2iIterator, "rnnType", defaultValue = "lstm")
    val dropoutProb = byLineFloatBuilder.build(x2iIterator, "dropoutProb")

    //
    // make the loadable parameters
    //
    val fwBuilder = mkBuilder(rnnType, numLayers, inputSize, rnnStateSize, parameters)
    val bwBuilder = mkBuilder(rnnType, numLayers, inputSize, rnnStateSize, parameters)

    new RnnLayer(parameters,
      inputSize, numLayers, rnnStateSize, useHighwayConnections,
      rnnType, fwBuilder, bwBuilder, dropoutProb)
  }

  private def mkBuilder(rnnType: String,
                        numLayers: Int,
                        inputSize: Int,
                        rnnStateSize: Int,
                        parameters: ParameterCollection): RnnBuilder = {
    rnnType match {
      case "gru" => new GruBuilder(numLayers, inputSize, rnnStateSize, parameters)
      case "lstm" => new LstmBuilder(numLayers, inputSize, rnnStateSize, parameters)
      case _ => throw new RuntimeException(s"""ERROR: unknown rnnType "$rnnType"!""")
    }
  }

  def initialize(config: Configured,
                 paramPrefix: String,
                 parameters: ParameterCollection,
                 inputSize: Int): Option[IntermediateLayer] = {
    if (!config.contains(paramPrefix)) {
      return None
    }

    val numLayers = config.getArgInt(paramPrefix + ".numLayers", Some(1))
    val rnnStateSize = config.getArgInt(paramPrefix + ".rnnStateSize", None)
    val useHighwayConnections = config.getArgBoolean(paramPrefix + ".useHighwayConnections", Some(false))
    val rnnType = config.getArgString(paramPrefix + ".type", Some("lstm"))
    val dropoutProb = config.getArgFloat(paramPrefix + ".dropoutProb", Some(RnnLayer.DEFAULT_DROPOUT_PROB))

    val wordFwRnnBuilder = mkBuilder(rnnType, numLayers, inputSize, rnnStateSize, parameters)
    val wordBwRnnBuilder = mkBuilder(rnnType, numLayers, inputSize, rnnStateSize, parameters)

    val layer = new RnnLayer(
      parameters,
      inputSize, numLayers, rnnStateSize, useHighwayConnections,
      rnnType, wordFwRnnBuilder, wordBwRnnBuilder,
      dropoutProb
    )

    Some(layer)
  }
}

File: main/src/main/python/pytorch/rnnLayer.py
from pytorch.intermediateLayer import IntermediateLayer
from pytorch.utils import *
import torch
import torch.nn as nn

class RnnLayer(IntermediateLayer):
    def __init__(self, 
        inputSize, 
        numLayers, 
        rnnStateSize, 
        useHighwayConnections, 
        rnnType, 
        wordRnnBuilder, 
        dropoutProb):
        super().__init__()
        self.inDim = self.inputSize = inputSize
        self.numLayers = numLayers
        self.rnnStateSize = rnnStateSize 
        self.useHighwayConnections = useHighwayConnections
        self.rnnType = rnnType
        self.wordRnnBuilder = mkBuilder(*wordRnnBuilder)
        self.dropoutProb = dropoutProb

        highwaySize = inputSize if useHighwayConnections else 0
        self.outDim = 2 * rnnStateSize + highwaySize

    def forward(self, inputExpressions, dropout):

        assert(inputExpressions is not None)

        States, _ = self.wordRnnBuilder(inputExpressions.unsqueeze(1))
        States = States.squeeze(1)
        if self.useHighwayConnections:
            States = torch.cat([States, inputExpressions], dim=1)

        return States

    def saveX2i(self):
        x2i = dict()
        x2i['inputSize'] = self.inputSize
        x2i['numLayers'] = self.numLayers
        x2i['rnnStateSize'] = self.rnnStateSize
        x2i['useHighwayConnections'] = 1 if self.useHighwayConnections else 0
        x2i['rnnType'] = self.rnnType
        x2i['dropoutProb'] = self.dropoutProb
        return x2i

    def __str__(self):
        return f"RnnLayer({self.rnnType}, {self.inDim}, {self.outDim})"

    @classmethod
    def load(cls, x2i):
        inputSize = x2i['inputSize']
        numLayers = x2i['numLayers']
        rnnType = x2i.get('rnnType', 'lstm')
        rnnStateSize = x2i['rnnStateSize']
        useHighwayConnections = x2i['useHighwayConnections'] == 1
        dropoutProb = x2i['dropoutProb']

        builder = (rnnType, numLayers, inputSize, rnnStateSize, dropoutProb)

        return cls(inputSize, numLayers, rnnStateSize, useHighwayConnections, rnnType, builder, dropoutProb)

    @classmethod
    def initialize(cls, config, paramPrefix, inputSize):

        if(not config.contains(paramPrefix)):
            return None

        numLayers = config.get_int(paramPrefix + ".numLayers", 1)
        rnnStateSize = config.get_int(paramPrefix + ".rnnStateSize", None)
        useHighwayConnections = config.get_bool(paramPrefix + '.useHighwayConnections', False)
        rnnType = config.get_string(paramPrefix + ".type", "lstm")
        dropoutProb = config.get_float(paramPrefix + ".dropoutProb", DEFAULT_DROPOUT_PROBABILITY)

        builder = (rnnType, numLayers, inputSize, rnnStateSize, dropoutProb)

        return cls(inputSize, numLayers, rnnStateSize, useHighwayConnections, rnnType, builder, dropoutProb)

def mkBuilder(rnnType, numLayers, inputSize, rnnStateSize, dropoutProb):
    if rnnType == 'gru':
        return nn.GRU(inputSize, rnnStateSize, numLayers, bidirectional=True, dropout=dropoutProb)
    elif rnnType == 'lstm':
        return nn.LSTM(inputSize, rnnStateSize, numLayers, bidirectional=True, dropout=dropoutProb)
    else:
        raise RuntimeError(f'ERROR: unknown rnnType "{rnnType}"!')





